### ISSUE TITLE
Prevent Delete Nodes and Edges

### ISSUE THREAD

**Issue by Vladislav747** [wbkd/react-flow#3092]
### Describe the Bug

I am using version 10.3.17 react-flow-renderer

i am trying to prevent delete with the Backspace btn. I need this function but i also need to make extra validation before delete. The only way i see now is to use onNodesChange handler and onEdgesChange handler to watch all elements that are need to be removed - but it's reallly not comfortable and not right. Is there way i can do smth to make extra validation before delete?

### Steps to Reproduce the Bug or Issue

- Try to delete node and to do smth before delete. 

### Expected behavior

- i have special handler - onBeforeDeleteElements
- delete happens only if i update nodes and edges intentionally by myself

### Platform

- OS: Linux
- Browser: Chrome

**Comment by moklick** [comment]
I like the idea of adding an `onBeforeDelete` handler. We will probably add it with the next major release.

**Comment by Vladislav747** [comment]
@moklick Hello! but is it possible to do onBeforeDeletle check somehow in 10.3.17 now? The problem i see is to prevent deletion(blinking of node) before any confirmation that i want to delete it.

**Comment by Vladislav747** [comment]
I have found answer to my question.

**Comment by stanislav-grin** [comment]
Hello @moklick, any chances this feature will appear in future releases?
I've implemented a custom delete with a button and I have a workflow after deleting a node, but users can still delete nodes by pressing Backspace and this is not desirable behavior, because it skips the post-delete workflow.
So either a `beforeDelete` handler to prevent deletion or a boolean flag like `deleteByBackspace` (default to `true`) would be very helpful.

**Comment by SejanasJoshua** [comment]
Can you please explain how to prevent delete of specific node types, need to validate before deleting, and also stop blinking of node.

**Comment by johannbuscail** [comment]
Hello ! Any update ?

**Comment by bcakmakoglu** [comment]
In v11, using the `onNodesChange` prop, you can validate if a node should be delete yourself.

```tsx
// only allow nodes that are *not* of type `custom` to be removed
function shouldNodeBeRemoved(node: Node) {
  if (node.type !== 'custom') {
    return true;
  }

  return false;
}

const BasicFlow = () => {
  const { getNode } = useReactFlow();
  const [nodes, , onNodesChange] = useNodesState(initialNodes);
  const [edges, setEdges, onEdgesChange] = useEdgesState(initialEdges);
  const onConnect = useCallback(
    (params: Edge | Connection) => setEdges((els) => addEdge(params, els)),
    [setEdges]
  );

  // wrap the `onNodesChange` function, so we can add our validation in there before a node gets removed
  function handleNodesChange(changes: NodeChange[]) {
    const nextChanges = changes.reduce((acc, change) => {
	  // if this change is supposed to remove a node we want to validate it first
      if (change.type === 'remove') {
        const node = getNode(change.id)

        // if the node can be removed, keep the change, otherwise we skip the change and keep the node
        if (shouldNodeBeRemoved(node)) {
          return [...acc, change];
        }
 
        // change is skipped, node is kept
        return acc;
      }

      // all other change types are just put into the next changes arr
      return [...acc, change];
    }, [] as NodeChange[])

	// apply the changes we kept
    onNodesChange(nextChanges);
  }

  return (
    <ReactFlow
      nodes={nodes}
      edges={edges}
      onNodesChange={handleNodesChange}
      onEdgesChange={onEdgesChange}
      onConnect={onConnect}
      nodeTypes={nodeTypes}
      fitView
    >
      <Background />
    </ReactFlow>
  );
};
```

[Here's an example sandbox.](https://codesandbox.io/p/sandbox/musing-bell-wfwnyz?file=%2Fsrc%2FFlow.tsx%3A44%2C1-92%2C3)

**Comment by dm-candeias** [comment]
@moklick Will you still add the "onBeforeDelete" feature? I found an issue with the presented workarround, i can prevent to delete the node but it will delete the edges connected to that node anyways.

**Comment by baibhavbista** [comment]
The problem with @bcakmakoglu 's solution using `onNodesChange` is that that callback is called only after the nodes&edges are deleted, so it has a high potential of being buggy (like @dm-candeias [elaborated above](https://github.com/xyflow/xyflow/issues/3092#issuecomment-1850591537))

In the absence of a `onBeforeDelete` callback, the way I ended up doing it is disabling React Flow's deleteKeyCode, adding custom key handler and then handling the logic there. See [the solution](https://github.com/xyflow/xyflow/discussions/2925#discussioncomment-6976130) by @AdamGerthel and [my elaboration](https://github.com/xyflow/xyflow/discussions/2925#discussioncomment-7907136)

cc @dm-candeias  in case you still need a solution

**Comment by bcakmakoglu** [comment]
@baibhavbista 
Not quite true. `onNodesChange` is not triggered *after* the nodes are actually deleted, it is triggered when a "changeset" proposing the removal of the nodes is sent to you at which point you *apply* those changes to your nodes and *then* they are deleted.
You basically interrupt the application of these changes for whatever validation etc. you want to do.

Using a custom key handler is completely fine though 😄

**Comment by baibhavbista** [comment]
Oh @bcakmakoglu sorry, my bad then
I had some weird issues when using that approach, maybe those were due to some other issue then?
Thanks for letting me know. I might try using that approach again!

**Comment by AndrewRayCode** [comment]
The `onNodesChange` answer works for me, but is this a hack, or is it intended to work this way? The [onNodesChange documentation](https://reactflow.dev/api-reference/react-flow#on-nodes-change) isn't clear, and also doesn't mention the ability to intercept / manually apply the changelist.

> Use this event handler to add interactivity to a controlled flow. It is called on node drag, select, and move.

The [useNodesState docs](https://reactflow.dev/api-reference/hooks/use-nodes-state#2) also don't clearly document this behavior:

> A handy callback that can take an array of NodeChanges and update the nodes state accordingly. You'll typically pass this directly to the onNodesChange prop of your <ReactFlow /> component

**Comment by kapxapot** [comment]
I am using the `onNodesChange` approach and it works for nodes. But the node's edges are still deleted.

I need to prevent specific edges' deletion in a specific case, but allow the same edges to be deleted directly or when other connected nodes are deleted (that are deletable). I don't see how it's possible with `onEdgesChange` since there is no info on the nodes' deletion context.

Let's say I have nodes A & B.

1. A should be not deletable.
2. B should be deletable.
3. When A is attempted to delete, A -> B edge should stay.
4. When A -> B edge is deleted directly, it should be deleted.
5. When B is attempted to delete, it should be deleted and A -> B to be deleted as well.

In other words when a user tries to delete node A, it should stay with all its edges in place. Otherwise, deletion must work as default.

How can I achieve that?

The only idea I have currently is to check if node A is selected currently and prevent its edges deletion in that case.

**Comment by GaMJun** [comment]
@kapxapot in my case i filtered the changes to only allow changes of type =  `position`, this will prevent the node to be selectable and removable (Im using Zustand)

```ts
  onNodesChange: (key) => (changes: NodeChange[]) => {
    const positionChanges = changes.filter((change) => change.type === "position");

    set((state) => ({
      nodes: applyNodeChanges(positionChanges, get().nodes),
    }));
  }
  ```

**Comment by kapxapot** [comment]
@GaMJun Thanks!

I managed to do it using the delete handler override and by tracking selected nodes & edges:
https://github.com/kapxapot/brightwood-builder/blob/master/src/components/flow.tsx#L203-L238

1. All selected edges are deleted.
2. If a node can be deleted, it's deleted with all its edges.
3. If a node can't be deleted, it and its edges don't change.

**Comment by manuesperon** [comment]
For anyone still trying to solve this issue, I managed to fix it by adding the  `nodesFocusable={false}` property to my `ReactFlow` component. Note that this will also prevent you from using other keyboard shortcuts like the arrow keys for moving the nodes for example.

**Comment by HoverBaum** [comment]
Nodes can be `deletable: false` now.

Hope this can help others to save some time 😅

There is a `deletable` property on nodes now. Set that to `deletable: false` when creating the node 🙃

**Comment by gitshot** [comment]
`onBeforeDelete={async () => false}
`
this worked for me...
